common/txmgr/address_state.go

```diff
@@ -0,0 +1,307 @@
+	txmgrtypes "github.com/smartcontractkit/chainlink/v2/common/txmgr/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	"gopkg.in/guregu/null.v4"
+)
+
+// AddressState is the state of a given from address
+type AddressState[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+] struct {
+	chainID     CHAIN_ID
+	fromAddress ADDR
+	txStore     PersistentTxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+
+	sync.RWMutex
+	idempotencyKeyToTx map[string]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	unstarted          *TxPriorityQueue[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]
+	inprogress         *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	// NOTE: currently the unconfirmed map's key is the transaction ID that is assigned via the postgres DB
+	unconfirmed map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+}
+
+// NewAddressState returns a new AddressState instance
+func NewAddressState[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+](
+	chainID CHAIN_ID,
+	fromAddress ADDR,
+	maxUnstarted int,
+	txStore PersistentTxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE],
+) (*AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE], error) {
+	as := AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		chainID:     chainID,
+		fromAddress: fromAddress,
+		txStore:     txStore,
+
+		idempotencyKeyToTx: map[string]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{},
+		unstarted:          NewTxPriorityQueue[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE](maxUnstarted),
+		inprogress:         nil,
+		unconfirmed:        map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{},
+	}
+
+	as.Lock()
+	defer as.Unlock()
+
+	// Load all unstarted transactions from persistent storage
+	offset := 0
+	limit := 50
+	for {
+		txs, count, err := txStore.UnstartedTransactions(offset, limit, as.fromAddress, as.chainID)
+		if err != nil {
+			return nil, fmt.Errorf("address_state: initialization: %w", err)
+		}
+		for i := 0; i < len(txs); i++ {
+			tx := txs[i]
+			as.unstarted.AddTx(&tx)
+			if tx.IdempotencyKey != nil {
+				as.idempotencyKeyToTx[*tx.IdempotencyKey] = &tx
+			}
+		}
+		if count <= offset+limit {
+			break
+		}
+		offset += limit
+	}
+
+	// Load all in progress transactions from persistent storage
+	ctx := context.Background()
+	tx, err := txStore.GetTxInProgress(ctx, as.fromAddress)
+	if err != nil {
+		return nil, fmt.Errorf("address_state: initialization: %w", err)
+	}
+	as.inprogress = tx
+	if tx != nil && tx.IdempotencyKey != nil {
+		as.idempotencyKeyToTx[*tx.IdempotencyKey] = tx
+	}
+
+	// Load all unconfirmed transactions from persistent storage
+	offset = 0
+	limit = 50
```

Comment:
Nit: Could make the load limit a global const. Would be easier to change in a single place if needed.